README.md
<h1 align="center"><img src="https://raw.githubusercontent.com/amphp/logo/master/repos/http-server.png?v=21-09-2018" alt="HTTP Server" width="350"></h1>

[![Build Status](https://travis-ci.org/amphp/http-server.svg?branch=master)](https://travis-ci.org/amphp/http-server)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/amphp/http-server/blob/master/LICENSE)

This package provides a non-blocking HTTP/1.1 and HTTP/2 application server written in PHP based on [Amp](https://github.com/amphp/amp).
Several features are provided in separate packages, such as the [WebSocket component](https://github.com/amphp/websocket-server).

The packages was previously named [`amphp/aerys`](https://github.com/amphp/aerys), but has been renamed to be easier to remember, as many people were having issues with the old name.

## Features

- [Static file serving](https://github.com/amphp/http-server-static-content)
- [WebSockets](https://github.com/amphp/websocket-server)
- [Dynamic app endpoint routing](https://github.com/amphp/http-server-router)
- Full TLS support
- Customizable GZIP compression
- HTTP/2.0 support
- Middleware hooks

## Requirements

- PHP 7

## Installation

```bash
composer require amphp/http-server
```

## Documentation

- [Official Documentation](http://amphp.org/http-server/)

## Example

```php
<?php

use Amp\Http\Server\RequestHandler\CallableRequestHandler;
use Amp\Http\Server\Server;
use Amp\Http\Server\Request;
use Amp\Http\Server\Response;
use Amp\Http\Status;
use Amp\Socket;
use Psr\Log\NullLogger;

// Run this script, then visit http://localhost:1337/ in your browser.

Amp\Loop::run(function () {
    $sockets = [
        Socket\listen("0.0.0.0:1337"),
        Socket\listen("[::]:1337"),
    ];
    
    $server = new Server($sockets, new CallableRequestHandler(function (Request $request) {
        return new Response(Status::OK, [
            "content-type" => "text/plain; charset=utf-8"
        ], "Hello, World!");
    }), new NullLogger);

    yield $server->start();

    // Stop the server gracefully when SIGINT is received.
    // This is technically optional, but it is best to call Server::stop().
    Amp\Loop::onSignal(SIGINT, function (string $watcherId) use ($server) {
        Amp\Loop::cancel($watcherId);
        yield $server->stop();
    });
});
```

```bash
php example.php
```

## Contributing

Please read [`CONTRIBUTING.md`](https://github.com/amphp/amp/blob/master/CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Security

If you discover any security related issues, please email [`contact@amphp.org`](mailto:contact@amphp.org) instead of using the issue tracker.

## License

The MIT License (MIT). Please see [LICENSE](./LICENSE) for more information.


CHANGELOG.md
### 0.8.3

 - Disabled compression if `ext-zlib` isn't available (#267)

### 0.8.2

- Fixed an issue when an HTTP/2 response is written immediately (#258).
- Performance recommendations are now logged as warnings when starting the server depending the mode set in options (debug or production), the value of the zend.assertions ini setting, and if the xdebug extension is loaded (related to #256).
- `Request::setBody()` and `Response::setBody()` now additionally accepts any value that can be cast to a string (such as integers, floats, and objects with a `__toString()` method) as the body content (#254).

### 0.8.1

- Fixed an issue where latency was increased dramatically on some systems compared to v0.7.x (#252).
- Fixed the `content-length` header being removed by `CompressionMiddleware` if the body was not long enough to be compressed.
- `ExceptionMiddleware` now writes the exception to the log to mimic the default behavior if it were not used.

### 0.8.0

This version is a major refactor, with many components being moved to separate libraries.

- Routing is now in [amphp/http-server-router](https://github.com/amphp/http-server-router)
- Static file serving (formerly `Root`) is now in [amphp/http-server-static-content](https://github.com/amphp/http-static-content)
- Form body parsing is now in [amphp/http-server-form-parser](https://github.com/amphp/http-server-form-parser)
- Multi-processing has been refactored to be general purpose and moved to [amphp/cluster](https://github.com/amphp/cluster)
- The WebSocket server component is now in [amphp/websocket-server](https://github.com/amphp/websocket-server)

A server is now created using an array of socket servers, an instance of `RequestHandler`, and a PSR-3 logger instance.

```php
use Amp\Http\Server\RequestHandler\CallableRequestHandler;
use Amp\Http\Server\Server;
use Psr\Log\NullLogger;

Amp\Loop::run(function () {
    $sockets = [
        Amp\Socket\listen("0.0.0.0:1337"),
        Amp\Socket\listen("[::]:1337"),
    ];
    
    $server = new Server($sockets, new CallableRequestHandler(function (Request $request) {
        return new Response(Status::OK, [
            "content-type" => "text/plain; charset=utf-8"
        ], "Hello, World!");
    }), new NullLogger);

    yield $server->start();

    // Stop the server gracefully when SIGINT is received.
    // This is technically optional, but it is best to call Server::stop().
    Amp\Loop::onSignal(SIGINT, function (string $watcherId) use ($server) {
        Amp\Loop::cancel($watcherId);
        yield $server->stop();
    });
});
```

Request handling has been changed significantly. Requests are now handled through interfaces modeled after [PSR-15](https://www.php-fig.org/psr/psr-15/), modified to be used in a non-blocking context.

Requests are handled by objects of classes implementing `RequestHandler`, which may be wrapped with any number of objects of classes implementing `Middleware`. Middleware is combined with a request handler using the `Middleware\stack()` function, returning an instance of `RequestHandler` that can be used with `Server` or anywhere else a `RequestHandler` may be used (such as routes in the routing library).

Please see the [documentation](https://amphp.org/http-server) for more information.

### 0.7.4

 - Fixed an issue where the timing of WebSocket writes could cause out of order messages.

### 0.7.3

 - Allow `amphp/file ^0.3`
 - Fixed sending pending responses when shutting down the server (#211)
 - More portable CPU count detection (#207)
 - WebSocket updates:
    - Reject websocket frames if RSV is not equal to `0` (will update as extensions are supported).
    - Accept zero-length frames starting a message.
    - Reject continuations if there's no started message.
    - Disable streaming for frames. A single frame is now always buffered. A message can still be streamed via multiple frames.

### 0.7.2

 - Fixed reading request port with HTTP/2

### 0.7.1

 - Fixed connection hangs if the process is forked while serving a request. See #182 and #192.

### 0.7.0

 - Fixed incorrect log level warning.
 - Fixed issue with referenced IPC socket blocking server shutdown.
 - Added support for unix sockets.
 - Added support for wildcard server names such as `localhost:*`, `*:80` and `*:*`.
 - Fixed buggy HTTP/1 pipelining.
 - Handle promises returned from generators in the config file correctly.
 - Added `-u` / `--user` command line option.
 - Correctly decode URL parameters with `urldecode()` instead of `rawurldecode()`.
 - Fixed freeze of websocket reading with very low `maxFramesPerSecond` and `maxBytesPerMinute`.
 - Removed `Router::__call()` magic, use `Router::route()` instead.

### 0.6.2

Retag of `v0.6.0`, as `v0.6.1` has been tagged wrongly, which should have been `v0.7.0`.

### 0.6.1

**Borked release.** Should have been `v0.7.0` and has been tagged as `v0.7.0` now.

### 0.6.0

Initial release based on Amp v2.

- Config files must return an instance of `Host` or an array of `Host` instances.
- `Aerys\Response::stream()` renamed to `write()` so that `Aerys\Response` may implement `Amp\ByteStream\OutputStream`.
- `Aerys\WebSocket\Endpoint::send()` split into three methods: `send()`, `broadcast()`, and `multicast()`.
- `Aerys\Body` removed. `Request::getBody()` returns an instance of `Amp\ByteStream\Message`.

### 0.5.0

_No information available._

### 0.4.7

_No information available._

### 0.4.6

_No information available._

### 0.4.5

_No information available._

### 0.4.4

_No information available._

### 0.4.3

 - Implemented monitoring system.
 - Always properly close HTTP/1.0 connections.
 - Fixed wildcard address matching.

### 0.4.2

_No information available._

### 0.4.1

 - Fixed caching issues in `Router` if multiple methods exist for specific URIs.
 
### 0.4.0

_No information available._

### 0.3.0

_No information available._

### 0.2.0

_No information available._

### 0.1.0

_No information available._


lib/App.php
<?php

namespace Aerys;

class App {
    const PORT = 'port';
    const ADDRESS = 'address';
    const NAME = 'name';
    const ENCRYPTION = 'encryption';
    const ORDER = 'order';
    const ROUTES = 'routes';
    const THREAD_ROUTES = 'thread_routes';
    const DOCUMENTS = 'documents';
    const WEBSOCKETS = 'websockets';
    const RESPONDERS = 'responders';

    private $port = 80;
    private $address = '*';
    private $name = '';
    private $standardRoutes = [];
    private $threadRoutes = [];
    private $websocketRoutes = [];
    private $documents = [];
    private $encryption = [];
    private $responders = [];
    private $order = [];

    /**
     * Define the host's port, IP and domain name
     *
     * Any valid port number [1-65535] may be used. Port numbers lower than 256 are reserved for
     * well-known services (like HTTP on port 80) and port numbers less than 1024 require root
     * access on UNIX-like systems. Port 80 is assumed in the absence of port specification. The
     * default port for encrypted sockets (https) is 443. If you plan to use encryption with this
     * app you'll generally want to use port 443.
     *
     * @param int $port The port number on which to listen
     * @param string $interface The IP address on which to bind this application
     * @param string $name The application domain name
     * @return \Aerys\App Returns the current object instance
     */
    public function setPort($port) {
        $this->port = $port;

        return $this;
    }

    /**
     * Define the IP interface on which the app will listen for requests
     *
     * The default wildcard IP value "*" translates to "all IPv4 interfaces" and is appropriate for
     * most scenarios. Valid values also include any IPv4 or IPv6 address. The string "[::]" denotes
     * an IPv6 wildcard.
     *
     * @param string $address The interface address (IP) on which the host is exposed
     * @return \Aerys\App Returns the current object instance
     */
    public function setAddress($address) {
        $this->address = $address;

        return $this;
    }

    /**
     * Define the app's host name (e.g. localhost or mysite.com or subdomain.mysite.com)
     *
     * A host name is only required if a server exposes more than one host.
     *
     * @param string $name
     * @return \Aerys\App Returns the current object instance
     */
    public function setName($name) {
        $this->name = $name;

        return $this;
    }

    /**
     * Define TLS encryption settings for this host
     *
     * The $tlsOptions array takes the following form:
     *
     * $tlsOptions = [
     *     'local_cert'             => '/path/to/mycert.pem', // *required
     *     'passphrase'             => 'mypassphrase',
     *     'allow_self_signed'      => TRUE,
     *     'verify_peer'            => FALSE,
     *     'ciphers'                => 'RC4-SHA:HIGH:!MD5:!aNULL:!EDH',
     *     'disable_compression'    => TRUE,
     *     'cafile'                 => NULL,
     *     'capath'                 => NULL
     * ];
     *
     * @param array $tlsOptions
     * @return \Aerys\App Returns the current object instance
     */
    public function setEncryption(array $tlsOptions) {
        $this->encryption = $tlsOptions;

        return $this;
    }

    /**
     * Specify an optional filesystem directory from which to serve static files
     *
     * The $options array takes the form:
     *
     * $options = [
     *     'indexes'                   => ['index.html', 'index.htm'],
     *     'eTagMode'                  => 'all',
     *     'expiresHeaderPeriod'       => 300,
     *     'defaultMimeType'           => 'text/plain',
     *     'customMimeTypes'           => [],
     *     'defaultTextCharset'        => 'utf-8',
     *     'cacheTtl'                  => 5,
     *     'memoryCacheMaxSize'        => 67108864,
     *     'memoryCacheMaxFileSize'    => 1048576
     * ];
     *
     * Note: websocket endpoint and dynamic HTTP route URIs always take precedence over filesystem
     * resources in the event of a routing conflict.
     *
     * @param string $rootDirectory
     * @param array $options An array specifying key-value options for static file serving
     * @return \Aerys\App Returns the current object instance
     */
    public function setDocumentRoot($rootDirectory, array $options = []) {
        $options['root'] = $rootDirectory;
        $this->documents = $options;

        return $this;
    }

    /**
     * Determine the order in which request responders are invoked for this application
     *
     * Valid values include:
     *
     * - App::WEBSOCKETS       (App::addWebsocket)
     * - App::ROUTES           (App::addRoute)
     * - App::RESPONDERS       (App::addResponder)
     * - App::DOCUMENTS        (App::setDocumentRoot)
     *
     * Any values specified that don't match the above list will result in a BootException
     * when the server is bootstrapped. Note that the above list is the default responder order.
     * User responders added via `App::addResponder` are always ordered internally by the
     * order in which they are added to the app.
     *
     * @param array $order
     * @return \Aerys\App Returns the current object instance
     */
    public function setResponderOrder(array $order) {
        $this->order = $order;

        return $this;
    }

    /**
     * Bind a non-blocking route handler for the specified HTTP method and URI path
     *
     * @param string $httpMethod The method for which this route applies
     * @param string $uriPath The route's URI path
     * @param mixed $handler Any callable or class::method construction string
     * @return \Aerys\App Returns the current object instance
     */
    public function addRoute($httpMethod, $uriPath, $handler) {
        $uriPath = '/' . ltrim($uriPath, '/');
        $this->standardRoutes[] = [$httpMethod, $uriPath, $handler];

        return $this;
    }

    /**
     * Bind a synchronous route handler to execute in a dedicated worker thread when matched
     *
     * @param string $httpMethod The method for which this route applies
     * @param string $uriPath The route's URI path
     * @param string $handler Any function or class::method construction string
     */
    public function addThreadRoute($httpMethod, $uriPath, $handler) {
        $uriPath = '/' . ltrim($uriPath, '/');
        $this->threadRoutes[] = [$httpMethod, $uriPath, $handler];

        return $this;
    }

    /**
     * Bind a websocket endpoint to the specified URI route
     *
     * Websocket routes are slightly different from other routes because they don't require an
     * HTTP method (GET is mandated by the protocol). Websocket routes require two arguments:
     *
     * - The websocket endpoint's URI path (routing regex allowed like all other routes)
     * - The name of the websocket endpoint class
     *
     * The third argument is an optional array specifying configuration values for this websocket
     * endpoint.
     *
     * @param string $uriPath The URI path on which to bind the endpoint
     * @param mixed $appClass A websocket endpoint class name
     * @param array $options An array specifying key-value options for this websocket endpoint
     * @return \Aerys\App Returns the current object instance
     */
    public function addWebsocketRoute($uriPath, $appClass, array $options = []) {
        $uriPath = '/' . ltrim($uriPath, '/');
        $this->websocketRoutes[] = [$uriPath, $appClass, $options];

        return $this;
    }

    /**
     * Add a user responder to the request-response chain
     *
     * User responders are always invoked in the order in which they are added to the App.
     *
     * @param mixed $responder Any callable or class::method construction string
     * @return \Aerys\App Returns the current object instance
     */
    public function addResponder($responder) {
        $this->responders[] = $responder;

        return $this;
    }

    /**
     * Retrieve an associative array summarizing the host definition
     *
     * @return array
     */
    public function toArray() {
        return [
            self::PORT          => $this->port,
            self::ADDRESS       => $this->address,
            self::NAME          => $this->name,
            self::ORDER         => $this->order,
            self::ENCRYPTION    => $this->encryption,
            self::DOCUMENTS     => $this->documents,
            self::ROUTES        => $this->standardRoutes,
            self::THREAD_ROUTES => $this->threadRoutes,
            self::WEBSOCKETS    => $this->websocketRoutes,
            self::RESPONDERS    => $this->responders,
        ];
    }
}
